tests/unit_tests/sparse_blas/source/sparse_trsv_buffer.cpp

```diff
@@ -0,0 +1,196 @@
+    }
+    catch (const sycl::exception &e) {
+        std::cout << "Caught synchronous SYCL exception during sparse TRSV:\n"
+                  << e.what() << std::endl;
+        print_error_code(e);
+        return 0;
+    }
+    catch (const oneapi::mkl::unimplemented &e) {
+        wait_and_free(main_queue, &handle);
+        return test_skipped;
+    }
+    catch (const std::runtime_error &error) {
+        std::cout << "Error raised during execution of sparse TRSV:\n" << error.what() << std::endl;
+        return 0;
+    }
+
+    // Compute reference.
+    prepare_reference_trsv_data(ia_host.data(), ja_host.data(), a_host.data(), m, int_index,
+                                uplo_val, transpose_val, diag_val, x_host.data(),
+                                y_ref_host.data());
+
+    // Compare the results of reference implementation and DPC++ implementation.
+    auto y_acc = y_buf.template get_host_access(sycl::read_only);
+    bool valid = check_equal_vector(y_acc, y_ref_host);
+
+    ev_release.wait_and_throw();
+    return static_cast<int>(valid);
+}
+
+class SparseTrsvBufferTests : public ::testing::TestWithParam<sycl::device *> {};
+
+/**
+ * Helper function to run tests in different configuration.
+ *
+ * @tparam fpType Complex or scalar, single or double precision type
+ * @param dev Device to test
+ * @param transpose_val Transpose value for the input matrix
+ */
+template <typename fpType>
+void test_helper(sycl::device *dev, oneapi::mkl::transpose transpose_val) {
+    double density_A_matrix = 0.8;
+    oneapi::mkl::index_base index_zero = oneapi::mkl::index_base::zero;
+    oneapi::mkl::uplo lower = oneapi::mkl::uplo::lower;
+    oneapi::mkl::diag nonunit = oneapi::mkl::diag::nonunit;
+    int m = 5;
+    bool use_optimize = true;
+
+    // Basic test
+    EXPECT_TRUEORSKIP(test<fpType>(dev, m, density_A_matrix, index_zero, lower, transpose_val,
+                                   nonunit, use_optimize));
+    // Test index_base 1
+    EXPECT_TRUEORSKIP(test<fpType>(dev, m, density_A_matrix, oneapi::mkl::index_base::one, lower,
+                                   transpose_val, nonunit, use_optimize));
+    // Test upper triangular matrix
+    EXPECT_TRUEORSKIP(test<fpType>(dev, m, density_A_matrix, index_zero, oneapi::mkl::uplo::upper,
+                                   transpose_val, nonunit, use_optimize));
+    // Test unit diagonal matrix
+    EXPECT_TRUEORSKIP(test<fpType>(dev, m, density_A_matrix, index_zero, lower, transpose_val,
+                                   oneapi::mkl::diag::unit, use_optimize));
+    // Test int64 indices
+    EXPECT_TRUEORSKIP(test<fpType>(dev, 15L, density_A_matrix, index_zero, lower, transpose_val,
+                                   nonunit, use_optimize));
+    if (!dev->is_cpu()) {
+        // Test without optimize_trsv
+        EXPECT_TRUEORSKIP(test<fpType>(dev, m, density_A_matrix, index_zero, lower, transpose_val,
+                                       nonunit, false));
+    }
```

Comment:
Provided you accept the other suggestion for conditional call to `shuffle_data()`, this should work for both CPU and GPU devices.
```suggestion
    // Test without optimize_trsv
    EXPECT_TRUEORSKIP(test<fpType>(dev, m, density_A_matrix, index_zero, lower, transpose_val,
                                       nonunit, false));
```

Reply:
Yes, that's better. It's fixed in https://github.com/oneapi-src/oneMKL/pull/407/commits/a762d79de03400ebf05b3ede85db6bb6242d9940